### ISSUE TITLE
How to include customized model in scoring a document?

### ISSUE THREAD

**Issue by namespace-Pt** [castorini/pyserini#890]
I have reproduced deep impact model, which consumes a passage and output the each token's score in the passage. I examined the tutorial about reproducing DeepImpact, however I didn't find a way to integrate my model's scoring into searching process. How can I achieve this?

**Comment by namespace-Pt** [comment]
I checked the downloaded corpus (json files) in the tutorial, and found each passage consists of three attributes: `id`, `content` and `vector`. In the `vector` field, I found a series of tokens and their scores that I expected to be the tokens' DeepImpact score. 

However, I found the score of each token is an interger, **does this mean a beforehand quantization?**. I also found the `content` field is not **expanded**, that is to say, some terms in the `vector` field are absent in the `content` field. I think this is a little weird since when retrieving only scores the overlapping tokens between the query and document are summed up, while some potentially overlapping tokens are missing from the document. 

In summary, two questions:

1. In case I've implemented my own DeepImpact model that produces each token's score and I want to use it for indexing and retrieving, can I just provide a similar json file containing `id`, `content` and `vector` fields?
2. Do I have to quantize the score of each token before generating the above json file?
3. Why are the passage's `content` field not expanded but the `vector` field comes from the expanded one?

**Comment by MXueguang** [comment]
Hi @namespace-Pt 
1. yes,  getting your custom `jsonl` file in the same format, you will be able to run the `pyserini.index` to index.
i.e. 
```
python -m pyserini.index -collection JsonVectorCollection \
 -input <directory containing your custom jsonl files> \
 -index <output directory> \
 -generator DefaultLuceneDocumentGenerator -impact -pretokenized \
 -threads 12
```
2. yes, the quantization needs to be done by yourself (i.e. when you prepare the `jsonl` file). We usually quantitative the weights into integers in range [0-255]. 
3. The contents are just an optional field here. Note that the weights are not generated from these `contents`. It was for the retrieval stage. (i.e. when we get a hit in retrieval, we are able to get the raw text of the original document.). This field can be left as empty.

**Comment by namespace-Pt** [comment]
Thank you very much Xueguang. I'll close the issue.

**Comment by lintool** [comment]
@namespace-Pt If you've done a successful reproduction, can you please send a PR to add to the log - at the bottom of the page?

**Comment by namespace-Pt** [comment]
@lintool I will do that after training and testing the model.

**Comment by namespace-Pt** [comment]
@lintool I've initiated a PR in `pyserini`.

**Comment by lintool** [comment]
Thanks, left a comment for you.
